flowmachine/tests/test_labelled_flows.py

```diff
@@ -0,0 +1,174 @@
+    ).join(
+        SubscriberHandsetCharacteristic(
+            "2016-01-01", "2016-01-03", characteristic="brand"
+        ),
+        "subscriber",
+        left_append="_hnd_type",
+        right_append="_brand",
+    )
+    return LabelledFlows(
+        loc1=loc_1,
+        loc2=loc_2,
+        labels=labels,
+        label_columns=["value_hnd_type", "value_brand"],
+    )
+
+
+def test_labelled_flow(labelled_flows, get_dataframe):
+
+    df = get_dataframe(labelled_flows)
+    assert len(df) == 349
+    # We lose some subscribers between locations
+    assert df.value.sum() == 490
+
+    labelled_outflow = labelled_flows.outflow()
+    out_df = get_dataframe(labelled_outflow)
+    assert all(out_df.columns == ["pcod_from", "value_label", "value"])
+
+
+def test_multiple_labels(get_dataframe, multi_labelled_flows):
+
+    assert multi_labelled_flows.out_label_columns == [
+        "value_hnd_type_label",
+        "value_brand_label",
+    ]
+    assert multi_labelled_flows.column_names == [
+        "pcod_from",
+        "pcod_to",
+        "value_hnd_type_label",
+        "value_brand_label",
+        "value",
+    ]
+    df = get_dataframe(multi_labelled_flows)
+    assert len(df) > 300
+    assert df.value.sum() == 490
+
+
+def test_spatial_unit_validation():
+    loc_1 = locate_subscribers(
+        "2016-01-01",
+        "2016-01-02",
+        spatial_unit=make_spatial_unit("admin", level=3),
+        method="most-common",
+    )
+
+    loc_2 = locate_subscribers(
+        "2016-01-02",
+        "2016-01-03",
+        spatial_unit=make_spatial_unit("admin", level=2),
+        method="most-common",
+    )
+
+    labels_1 = SubscriberHandsetCharacteristic(
+        "2016-01-01", "2016-01-03", characteristic="hnd_type"
+    )
+
+    with pytest.raises(ValueError, match="same spatial unit"):
+        _ = LabelledFlows(loc1=loc_1, loc2=loc_2, labels=labels_1)
+
+
+def test_column_validation():
+    loc_1 = locate_subscribers(
+        "2016-01-01",
+        "2016-01-02",
+        spatial_unit=make_spatial_unit("admin", level=3),
+        method="most-common",
+    )
+
+    loc_2 = locate_subscribers(
+        "2016-01-02",
+        "2016-01-03",
+        spatial_unit=make_spatial_unit("admin", level=3),
+        method="most-common",
+    )
+
+    labels_1 = SubscriberHandsetCharacteristic(
+        "2016-01-01", "2016-01-03", characteristic="hnd_type"
+    )
+    with pytest.raises(ValueError, match="not present in"):
+        _ = LabelledFlows(
+            loc1=loc_1, loc2=loc_2, labels=labels_1, label_columns=["notavalue"]
+        )
+
+
+def test_geojson(labelled_flows):
+
+    out = labelled_flows.to_geojson_string()
+    dict = loads(out)
+    assert all(
+        test in dict["features"][0]["properties"]["outflows"].keys()
+        for test in ["Smart", "Feature"]
+    )
+    assert all(
+        test in dict["features"][0]["properties"]["inflows"].keys()
+        for test in ["Smart", "Feature"]
+    )
```

Comment:
Would be good to check that the locations are nested within the labels here (e.g. by checking for a particular pcod in `dict["features"][0]["properties"]["inflows"]["Feature"].keys()`)